DOC for refactored compression (GH14576) + BUG: bz2-compressed URL with C engine (GH14874)

dhimmel opened the pull request:
Follow up on https://github.com/pandas-dev/pandas/pull/14576, which refactored compression code to expand URL support.

Fixes up some small remaining issues and adds a what's new entry.

- [x] Closes #14874

dhimmel committed:
DOC: Improve _infer_compression docstring

https://github.com/pandas-dev/pandas/commit/4a5aec40e8b2d6789f946e3e5b5b07ba5e753eb6#commitcomment-20178761

dhimmel committed:
ENH: Support bz2 compression in PY2 for c engine

Closes https://github.com/pandas-dev/pandas/issues/14874

dhimmel committed:
TST: Read compressed URLs with c engine

dhimmel committed:
DOC: What's New for refactored compression code

Add what's new corresponding to https://github.com/pandas-dev/pandas/pull/14576.

jorisvandenbossche commented on doc/source/whatsnew/v0.20.0.txt:
```diff
@@ -44,6 +44,18 @@ fixed-width text files, and :func:`read_excel` for parsing Excel files.
+   url = ('https://github.com/pandas-dev/pandas/raw/master/' +
+          'pandas/io/tests/parser/data/salaries.csv.bz2')
+   pd.read_table(url, compression='infer')  # default, infer compression
+   pd.read_table(url, compression='xz')  # explicitly specify compression
```
This will give an error, as the url is bz2, is that the intention?

dhimmel replied:
No, this is a mistake... they should both be bz2 or xz.

jorisvandenbossche replied:
Can you also add `.head()` or so, as otherwise the output will get a bit long for just demonstrating the compression keyword

jorisvandenbossche replied:
or assign to `df = ` without showing its content, also a possibility

dhimmel replied:
Didn't realize this code would get executed. Now assigns to `df` and then uses head to display first two rows. In https://github.com/pandas-dev/pandas/pull/14880/commits/cb40a41bd26ef50aedb3923235d648cd4cd4aa94.

jorisvandenbossche commented on pandas/parser.pyx:
```diff
@@ -621,8 +621,10 @@ cdef class TextReader:
-                                     'handle')
+                    content = source.read()
+                    source.close()
+                    from pandas.compat import StringIO
```
`pandas.compat` is already imported as `compat`

dhimmel replied:
Fixed those two issues. Was a nice opportunity for me to [try out `--fixup`](http://fle.github.io/git-tip-keep-your-branch-clean-with-fixup-and-autosquash.html). We can do the `--autosquash` when the PR looks ready to go.

dhimmel committed:
DOC: Reword get_filepath_or_buffer docstring

dhimmel committed:
TST: check bz2 compression in PY2 c engine

jreback commented on doc/source/whatsnew/v0.20.0.txt:
```diff
@@ -44,6 +44,19 @@ fixed-width text files, and :func:`read_excel` for parsing Excel files.
    pd.read_fwf(StringIO(data)).dtypes
    pd.read_fwf(StringIO(data), dtype={'a':'float64', 'b':'object'}).dtypes
 
+Reading dataframes from URLs, in :func:`read_csv` or :func:`read_table`, now
```
I would make a sub-section

jreback commented on doc/source/whatsnew/v0.20.0.txt:
```diff
@@ -44,6 +44,19 @@ fixed-width text files, and :func:`read_excel` for parsing Excel files.
+Reading dataframes from URLs, in :func:`read_csv` or :func:`read_table`, now
+supports additional compression methods (`xz`, `bz2`, `zip`). Previously, only
+`gzip` compression was supported. By default, compression of URLs and paths are
+now both inferred using their file extensions.
```
can you add all of the issues that were closed here, e.g. ``(:issue:`....`, :issue:`...`)``

jreback commented:
lgtm (minor doc comment)

dhimmel committed:
DOC: Improve what's new

Reference corresponding issues in What's New.

Change code example to use string formating for improved modularity.

Add what's new id

dhimmel commented:
Okay I cleaned up the commit history with

```sh
git rebase --interactive --autosquash 30025d82564fc27fbab58fbd791009e5b77a23db
git push --force
```

Assuming tests pass, I think this PR is ready. @jreback / @jorisvandenbossche -- you may want to double check the what's new, since made some additional changes in https://github.com/pandas-dev/pandas/pull/14880/commits/3f4cd4518bcbe70758faa5e0c86d96e56bcb8e30.

jorisvandenbossche commented on pandas/io/common.py:
```diff
@@ -247,23 +247,26 @@ def file_path_to_url(path):
-    Get file handle for given path/buffer and mode.
+    Get the compression method for filepath_or_buffer. If compression='infer',
+    the inferred compression method is returned. Otherwise, the input
+    compression method is returned unchanged, unless it's invalid, in which case
```
PEP8 line too long (see bottom third travis test)

jorisvandenbossche reviewed (commented):
Only some remaining lint:

pandas/io/tests/parser/compression.py:11:1: F401 'pandas.compat' imported but unused

dhimmel committed:
STY: PEP8 fixes

dhimmel commented:
@jorisvandenbossche, I fixed the PEP8 issues. Had to rebase to resolve a conflict that arose in `doc/source/whatsnew/v0.20.0.txt` and did some squashing/rearranging. As a result, the chronological commit order [GitHub shows](https://help.github.com/articles/why-are-my-commits-in-the-wrong-order/), isn't the actual commit order.

Also there [may be an issue](https://travis-ci.org/pandas-dev/pandas/jobs/184064359) with S3 tests failing. I was going to wait for the latest build and see if the issues are still there.

jorisvandenbossche commented:
It's indeed failing. What is the reason for this? It seems you didn't change anything related to that?

> did some squashing/rearranging

Just as a personal recommendation, I would not put too much effort in it. Previously we asked people to squash their PRs, but now the github interface (and our own tooling) has improved, and we squash on merge. Your previous PR was just a special case as there were two commit authors to retain.

dhimmel committed:
TST: Read bz2 files from S3 in PY2

Addresses https://github.com/pandas-dev/pandas/issues/14874

dhimmel commented:
> It's indeed failing. What is the reason for this? It seems you didn't change anything related to that?

Some S3 tests were still expecting bz2 failure in Python 2. Fixed in https://github.com/pandas-dev/pandas/pull/14880/commits/8568aedc58cd836390e8eddff7e844b0bb9ea875.

> Previously we asked people to squash their PRs, but now the github interface (and our own tooling) has improved, and we squash on merge.

Instead of rebasing is it acceptable to merge in the current master? This way we don't keep destroying references to old commit hashes.

jreback commented on doc/source/whatsnew/v0.20.0.txt:
```diff
@@ -64,6 +64,27 @@ Strings passed to ``DataFrame.groupby()`` as the ``by`` parameter may now refere
+dataframes from URLs in :func:`read_csv` or :func:`read_table` now supports
+additional compression methods: ``xz``, ``bz2``, and ``zip`` (:issue:`14570`).
+Previously, only ``gzip`` compression was supported. By default, compression of
+URLs and paths are now both inferred using their file extensions. Additionally,
```
``The compression code``

``paths are not inferred using`` (remove both)

Additionally, support for bz2 compress in the python 2 c-engine improved.

dhimmel replied:
Addressed comments 1 and 3 in https://github.com/pandas-dev/pandas/pull/14880/commits/e1b5d4200fdb4b9f4f7fb590072ead355e9b1517. @jreback, I didn't change:

> By default, compression of URLs and paths are now both inferred using their file extensions.

Previously, compression of paths was by default inferred from their extension, but not URLs. Now both are inferred by their extension. Am I missing something?

jreback commented on doc/source/whatsnew/v0.20.0.txt:
```diff
@@ -64,6 +64,27 @@ Strings passed to ``DataFrame.groupby()`` as the ``by`` parameter may now refere
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Compression code was refactored (:issue:`12688`). As a result, reading
+dataframes from URLs in :func:`read_csv` or :func:`read_table` now supports
```
if there are any other issues that were closed by this, pls list them as well.

dhimmel replied:
Rechecked... they're all already listed.

jreback commented:
@dhimmel very minor doc changes. ping on when green.

@TomAugspurger after this is merged, pls rebase #13137

dhimmel committed:
Address what's new review comments

dhimmel commented:
> @dhimmel very minor doc changes. ping on when green.

@jreback @jorisvandenbossche -- we're green.

jreback commented:
thanks @dhimmel great responsiveness!
